File: completed/2092.go
package main

import (
    "fmt"
    "sort"
)

func findAllPeople(n int, meetings [][]int, firstPerson int) []int {
    parents := make([]int, n)
    for i := range parents {
        parents[i] = i
    }
    parents[firstPerson] = 0

    var find func(person int) int
    find = func(person int) int {
        if parents[person] != person {
            parent := find(parents[person])
            parents[person] = parent
        }
        return parents[person]
    }
    var union = func(parent int, child int) {
        parents[find(child)] = find(parent)
    }
    var reset = func(person int) {
        parents[person] = person
    }

    sort.Slice(meetings, func(i int, j int) bool {
        return meetings[i][2] < meetings[j][2];
    })
    for i := 0; i < len(meetings); i++ {
        same := make(map[int]bool)

        j := i
        for j < len(meetings) && meetings[i][2] == meetings[j][2] {
            same[meetings[j][0]] = true
            same[meetings[j][1]] = true
            union(meetings[j][0], meetings[j][1])
            j++
        }
        i = j - 1;

        target := find(0)
        for person := range same {
            if find(person) != target {
                reset(person)
            }
        }
    }

    target := find(0)
    know := make([]int, 0)
    for i := range parents {
        if find(i) == target {
            know = append(know, i)
        }
    }
    return know
}

func main() {
    n := 11
    meetings := [][]int{ { 5, 1, 4 }, { 0, 4, 18 } }
    firstPerson := 1
    fmt.Println(findAllPeople(n, meetings, firstPerson))
}

File: completed/2092.java
class Solution {
    private void dfs(
        int index,
        Map<Integer, List<Integer>> adjacencyLists,
        Set<Integer> visited,
        List<Integer> component
    ) {
        // 1. Recursively construct the component
        component.add(index);
        visited.add(index);

        for (int adjacent : adjacencyLists.get(index)) {
            if (visited.contains(adjacent)) {
                // already visited
                continue;
            }

            dfs(adjacent, adjacencyLists, visited, component);
        }
    }

    private List<List<Integer>> getConnectedComponents(
        int[][] meetings,
        int left,
        int right
    ) {
        // 1. Construct adjacency lists for easy look up
        Map<Integer, List<Integer>> adjacencyLists = new HashMap<>();
        for (int i = left; i < right; i++) {
            int x = meetings[i][0];
            int y = meetings[i][1];
            adjacencyLists.computeIfAbsent(x, _ -> new ArrayList<>()).add(y);
            adjacencyLists.computeIfAbsent(y, _ -> new ArrayList<>()).add(x);
        }

        // 2. DFS from each person and construct the components
        List<List<Integer>> components = new ArrayList<>();
        Set<Integer> visited = new HashSet<>();
        for (int source : adjacencyLists.keySet()) {
            if (visited.contains(source)) {
                // already visited
                continue;
            }

            List<Integer> component = new ArrayList<>();
            dfs(source, adjacencyLists, visited, component);
            components.add(component);
        }
        return components;
    }

    public List<Integer> findAllPeople(int n, int[][] meetings, int firstPerson) {
        // 1. Order meetings by time
        Arrays.sort(meetings, (m1, m2) -> m1[2] - m2[2]);

        // 2. Keep track of who knows
        boolean[] knows = new boolean[n];
        knows[0] = true;
        knows[firstPerson] = true;

        // 3. Go through all meetings and update who knows
        int left = 0;
        while (left < meetings.length) {
            // 4. Get all meetings at the same time
            List<Integer[]> concurrentMeetings = new ArrayList<>();

            int time = meetings[left][2];
            int right = left;
            while (right < meetings.length) {
                if (meetings[right][2] > time) {
                    break;
                }
                right++;
            }

            // 5. Find the connected components
            List<List<Integer>> components = getConnectedComponents(meetings, left, right);

            // 6. Update everyone in the component if needed
            for (List<Integer> component : components) {
                boolean someoneKnows = false;
                for (int person : component) {
                    if (knows[person]) {
                        someoneKnows = true;
                        break;
                    }
                }

                if (someoneKnows) {
                    for (int person : component) {
                        knows[person] = true;
                    }
                }
            }

            left = right;
        }

        // 7. Construct result
        List<Integer> result = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            if (knows[i]) {
                result.add(i);
            }
        }
        return result;
    }
}

